keps/sig-scheduling/5027-dra-admin-controlled-device-attributes/README.md

```diff
@@ -0,0 +1,772 @@
+
+Creating a ResourceSlicePatch is racing with on-going scheduling attempts,
+which is unavoidable. Removing a device from a ResourceSlice has the same
+problem.
+
+### Risks and Mitigations
+
+From a security perspective, permission to patch device attributes is
+expected to be limited to privileged users who already have the ability to add
+or remove DRA drivers, so there won't be a substantial difference.
+
+Performance in the scheduler could be an issue. This will be mitigated by
+caching the patched devices and
+(re-)applying patches only when they or the device definitions change, which
+should be rare.
+
+## Design Details
+
+### API
+
+The ResourceSlicePatch is a cluster-scoped type in the `resource.k8s.io` API
+group, initially in `v1alpha3` (the alpha version in Kubernetes 1.32). Because
+it may be useful to clean up after disabling the feature and because the
+device taint feature also uses this type, it gets served unconditionally as long as
+the `v1alpha3` version is enabled. Fields related specifically to this KEP
+are feature-gated.
+
+```Go
+type ResourceSlicePatch struct {
+metav1.TypeMeta
+    // Standard object metadata
+    // +optional
+    metav1.ObjectMeta
+
+    // Changing the spec automatically increments the metadata.generation number.
+    Spec ResourceSlicePatchSpec
+}
+
+type ResourceSlicePatchSpec struct {
+    // Devices defines how to patch device attributes and taints.
+    Devices DevicePatch
+}
+
+// DevicePatch selects one or more devices by class, driver, pool, device names
+// and/or CEL selectors. All of these criteria must be satisfied by a device, otherwise
+// it is ignored by the patch. A DevicePatch with no selection criteria is
+// valid and matches all devices.
+type DevicePatch struct {
+    // Filter defines which device(s) the patch is applied to.
+    //
+    // +optional
+    Filter *DevicePatchFilter
+
+    // If a ResourceSlice and a DevicePatch define the same attribute or
+    // capacity, the value of the DevicePatch is used. If multiple
+    // different DevicePatches match the same device, then the one with
+    // the highest priority wins. If the priorities are the same, it is non-deterministic
+    // which patch is used.
+    //
+    // +optional
+    Priority *int
+
+    // Attributes defines the set of attributes to patch for matching devices.
+    // The name of each attribute must be unique in that set and
+    // include the domain prefix.
+    //
+    // In contrast to attributes in a ResourceSlice, entries here are allowed to
+    // be marked as empty by setting their null field. Such entries remove the
+    // corresponding attribute in a ResourceSlice, if there is one, instead of
+    // overriding it. Because entries get removed and are not allowed in
+    // slices, CEL expressions do not need need to deal with null values.
+    //
+    // The maximum number of attributes and capacities in the DevicePatch combined is 32.
```

Comment:
Just curious: why the maximum number is 32 here? Is there any reason to use 32 instead of other numbers?

Reply:
It's the same as for a ResourceSlice entry. We could make it larger here because we are less constrained by overall object size, but that seems unnecessary.